pkg/scheduler/algorithm/priorities/even_pods_spread.go

```diff
@@ -0,0 +1,220 @@
+
+	"k8s.io/klog"
+)
+
+type topologyPair struct {
+	key   string
+	value string
+}
+
+type topologySpreadConstraintsMap struct {
+	// nodeNameToPodCounts is keyed with node name, and valued with the number of matching pods.
+	nodeNameToPodCounts map[string]int64
+	// topologyPairToPodCounts is keyed with topologyPair, and valued with the number of matching pods.
+	topologyPairToPodCounts map[topologyPair]*int64
+}
+
+func newTopologySpreadConstraintsMap(len int) *topologySpreadConstraintsMap {
+	return &topologySpreadConstraintsMap{
+		nodeNameToPodCounts:     make(map[string]int64, len),
+		topologyPairToPodCounts: make(map[topologyPair]*int64),
+	}
+}
+
+// Note: the <nodes> passed in are the "filtered" nodes which have passed Predicates.
+// This function iterates <nodes> to filter out the nodes which don't have required topologyKey(s),
+// and initialize two maps:
+// 1) t.topologyPairToPodCounts: keyed with both eligible topology pair and node names.
+// 2) t.nodeNameToPodCounts: keyed with node name, and valued with a *int64 pointer for eligible node only.
+func (t *topologySpreadConstraintsMap) initialize(pod *v1.Pod, nodes []*v1.Node) {
+	constraints := getSoftTopologySpreadConstraints(pod)
+	for _, node := range nodes {
+		match := true
+		var pairs []topologyPair
+		for _, constraint := range constraints {
+			tpKey := constraint.TopologyKey
+			if _, ok := node.Labels[tpKey]; !ok {
+				// Current node isn't qualified for the soft constraints,
+				// so break here and the node will hold default value (nil).
+				match = false
+				break
+			}
+			pairs = append(pairs, topologyPair{key: tpKey, value: node.Labels[tpKey]})
+		}
+		if match {
+			for _, pair := range pairs {
+				if t.topologyPairToPodCounts[pair] == nil {
+					t.topologyPairToPodCounts[pair] = new(int64)
+				}
+			}
+		} else {
+			// For those nodes which don't have all required topologyKeys present, it's intentional to
+			// assign nodeNameToPodCounts[nodeName] as -1, so that we're able to score these nodes to 0 afterwards.
+			t.nodeNameToPodCounts[node.Name] = -1
+		}
+	}
+}
+
+// CalculateEvenPodsSpreadPriority computes a score by checking through the topologySpreadConstraints
+// that are with WhenUnsatisfiable=ScheduleAnyway (a.k.a soft constraint).
+// The function works as below:
+// 1) In all nodes, calculate the number of pods which match <pod>'s soft topology spread constraints.
+// 2) Sum up the number to each node in <nodes> which has corresponding topologyPair present.
+// 3) Finally normalize the number to 0~10. The node with the highest score is the most preferred.
+// Note: Symmetry is not applicable. We only weigh how incomingPod matches existingPod.
+// Whether existingPod matches incomingPod doesn't contribute to the final score.
+// This is different with the Affinity API.
+func CalculateEvenPodsSpreadPriority(pod *v1.Pod, nodeNameToInfo map[string]*schedulernodeinfo.NodeInfo, nodes []*v1.Node) (schedulerapi.HostPriorityList, error) {
```

Comment:
Where do we call this priority function? This is the only place "CalculateEvenPodsSpreadPriority" string appears.

Reply:
Good question.. I realized the issue only until I wrote the integration test (i.e. in PR6, you can see the fix - it's actually an typo in `defaults.go`...)

But as you raised this, to avoid confusion, let me move that fix in PR6 to this PR.

BTW: we can see the benefits of having different kinds of tests :)

Reply:
Oh... In addition to that typo, I made another rebasing mistake..

Reply:
Done. Squashed to commit "EvenPodsSpread: Define a new Priority".

Thanks for the catch! (Otherwise, the next PR is the only guard...)

To get the benefits of making each PR CI green, I paid way too much energy on rebasing...

Reply:
> Done. Squashed to commit "EvenPodsSpread: Define a new Priority".
> 
> Thanks for the catch! (Otherwise, the next PR is the only guard...)
> 
> To get the benefits of making each PR CI green, I paid way too much energy on rebasing...

The rebasing could cause tons of work. Thanks for breaking this into pieces.